pkg/drivers/sqlite/sqlite.go

```diff
@@ -77,172 +65,68 @@ func NewVariant(ctx context.Context, driverName, dataSourceName string) (server.
-// the given new ones.
-func alterIndices(txn *sql.Tx, d *generic.Generic) error {
-	if d.LockWrites {
-		d.Lock()
-		defer d.Unlock()
-	}
-
-	dropIndices := []string{
-		`DROP INDEX IF EXISTS kine_name_index`,
-		`DROP INDEX IF EXISTS kine_name_prev_revision_uindex`,
-	}
-
-	// Drop the old indices
-	for _, stmt := range dropIndices {
-		_, err := txn.Exec(stmt)
-		if err != nil {
-			return err
-		}
-	}
-
-	createIndices := []string{
-		`CREATE INDEX IF NOT EXISTS kine_name_index ON kine (name, id)`,
-		`CREATE UNIQUE INDEX IF NOT EXISTS kine_name_prev_revision_uindex ON kine (prev_revision, name)`,
-	}
-
-	// Create the new indices
-	for _, stmt := range createIndices {
-		_, err := txn.Exec(stmt)
-		if err != nil {
-			return err
+	for i := 0; i < retryAttempts; i++ {
+		if err := migrate(ctx, txn); err != nil {
+			if sqliteErr, ok := err.(sqlite3.Error); ok && sqliteErr.ExtendedCode == sqlite3.ErrBusySnapshot {
+				// Another write transaction got in the way (SERIALIZABLE isolation).
+				// The whole transaction can be retried.
+				continue
+			}
+			return errors.Wrap(err, "migration failed")
 		}
+		break
 	}
 
-	return nil
-}
-
-// countTable counts the number of tables with the given name, in the given transaction.
-func hasTable(ctx context.Context, txn *sql.Tx, tableName string) (bool, error) {
-	tableListSQL := fmt.Sprintf(`SELECT COUNT(*) FROM sqlite_master WHERE type = 'table' AND name = '%s'`, tableName)
-	row := txn.QueryRowContext(ctx, tableListSQL)
-	var tableCount int
-	if err := row.Scan(&tableCount); err != nil {
-		return false, err
-	}
-
-	return tableCount != 0, nil
+	return txn.Commit()
 }
 
-// migration copies rows of the old key value table to the kine
-// table ONLY IF the key value table exists and migration has not been
-// done already. It creates the kine table if it doesn't exist.
-func migration(ctx context.Context, txn *sql.Tx, d *generic.Generic) error {
-	userVersionSQL := `PRAGMA user_version`
-	row := txn.QueryRowContext(ctx, userVersionSQL)
-
+// migrate tries to migrate from a version of the database
+// to the target one.
+func migrate(ctx context.Context, txn *sql.Tx) error {
 	var userVersion int
+
+	row := txn.QueryRowContext(ctx, `PRAGMA user_version`)
 	if err := row.Scan(&userVersion); err != nil {
 		return err
 	}
-	// No need for migration - marker has already been set
-	if userVersion == databaseSchemaVersion {
-		return nil
-	}
 
-	if userVersion > databaseSchemaVersion {
-		return errors.Errorf("unsupported version: %d", userVersion)
-	}
-
-	kineTableExists, err := hasTable(ctx, txn, "kine")
-	if err != nil {
-		return err
-	}
-	// Create Kine table if it doesn't already exist
-	if !kineTableExists {
-		if err := createKineTable(txn); err != nil {
+	switch userVersion {
+	case 0:
+		if err := applySchemaV1(ctx, txn); err != nil {
 			return err
 		}
+		fallthrough
+	case databaseSchemaVersion:
+		break
+	default:
+		// FIXME this needs better handling
```

Comment:
I wonder if we could handle it this way:
Define a constant/variable int somehow to represent the latest version,
If the `kine` table already exists, then check its version, and if it's not equal to the latest version:
- create new table with the latest schema,
- copy over older kine table's contents to the new table.
- set the version to the latest version

So the `databaseSchemaVersion` (==1) would represent the earliest version with a `kine` table in place, and `latestSchemaVersion` would be the latest at the time. Maybe a programmatic way to pass the latest schema as well, and feed that to the migrate method, i.e. (`newSchema string`).

Maybe for a later rewrite of Kine...